### ISSUE TITLE
'Has no TestFixtures' Error could be improved when nunit.framework.dll is not the referenced version

### ISSUE THREAD

**Issue by jimalad** [nunit/nunit#2806]
'Has no TestFixtures' really threw me off the scent when our CI build broke yesterday. Someone had upgraded nunit3 (nunit.framework.dll) in one of the test projects in our solution.

Like others getting stuck with this error, our build script compiles all solution project bin outputs to a single bin output directory (msbuild OutDir). I'm not at all sure but I'd guess that the project build order might mean that the last test project built will produce the 'winning' version of nunit.framework.dll. In my case, it was too new for nunit3-console.exe on our build server and I got a broken build on a unit test project that I hadn't changed and an error that I struggled to understand.

Would it be possible to improve this nunit3-console.exe error so it instead says something like...
1) Invalid : c:\what\ever\MyAssembly.Tests.dll
Has no TestFixtures. nunit.framework.dll (3.10.1) is incompatible with nunit3-console.exe (3.8.1)"

references...
https://stackoverflow.com/a/47290562/9050451
#1207

**Comment by ChrisMaddock** [comment]
> In my case, it was too new for nunit3-console.exe on our build server

This shouldn't be the case - any 3.x version of the framework should run on any 3.x version of the console.

There are two potential errors I see here:
1) There's a bug in NUnit where a specific version of the framework doesn't work with a specific version of the console. This should be fixed, if possible!
2) The issue was instead caused by one of your test projects expecting to reference a certain version of the NUnit framework, but finding another in the bin directory.

Could you investigate further, and let us know which is the case? If it's the former - please let us know the versions.

**Comment by jimalad** [comment]
Thank you for the cue investigate further.

I've just created two nunit3 projects. One built against 3.8.1 and one against 3.10.1 and I build them independently to their own bin directories. As you say, console 3.8 is happy to run tests from either assembly, so sorry for being completely wrong about that!

To reproduce 'Has no TestFixtures' I replaced nunit.framework.dll (in the bin output directory of the 3.10.1 project) with nunit.framework.dll (3.8.1) and retried running the console.

So that's confirmed that it's point 2 in your list.

So is this therefore happening because we're using a bad workflow? The msbuild OutDir is maybe bad, or because we haven't worked out if nuget can enforce peg all nunit references on all projects to the same version across the solution?

Should nunit-console3.exe be checking and erroring if the test assembly it's running has bound to the wrong version of the framework dll from the one that the test assembly was originally built with?

**Comment by rprouse** [comment]
> So is this therefore happening because we're using a bad workflow?

Yes, although we should probably give a better error message because this happens often enough. Whenever you build all of your projects to one bin directory, you should always ensure that all nuget packages for all projects are at the same version, not just NUnit. If you don't do that, you will get random versions of assemblies in your output folder depending on build order and even what was recently built.

Visual Studio has a tab in the NuGet manager to help with this,

1. Right click on your solution and select Manage NuGet Packages for Solution
2. Click the Consolidate tab
3. This tab shows you NuGet packages that are different between projects and allows you to update them to be the same.

![image](https://user-images.githubusercontent.com/493828/38744269-53cfeb9a-3f0f-11e8-9cb4-63c5745dcef5.png)

**Comment by rprouse** [comment]
I've updated the title of this issue and am making it an enhancement. If we fail to load a test assembly because the version of the framework is not correct, we should present a message that helps users diagnose the problem.

**Comment by jimalad** [comment]
Thank you all for your help with the diagnosis - I'm most grateful. Thanks also for the very clear explaination how NuGet Consolidate helps to avoid the problem.
I was going to suggest changing the issue title too.
